### PR TITLE
Setup a monorepo with a JupyterLab extension

### PR DESCRIPTION
**Pull request by hbcarlos**
# TODO:
- [x] Clean up packages
- [x] Fix/update GitHub action
- [x] Setup Lerna
- [x] Create a lab extension
- [x] Bring in the remaining plugins
  - [x] Register YDrive
  - [x] File browser with YDrive

# Problems
#### Version
* JavaScript and Python packages have different versions

# TODO in a follow-up PR
- [ ] Add documentation
  - [ ] Bring RTC docs from JupyterLab
    * https://github.com/jupyterlab/jupyterlab/blob/master/docs/source/extension/documents.rst#shared-models ?
    * https://github.com/jupyterlab/jupyterlab/blob/master/docs/source/user/rtc.rst ?
- [ ] Add tests
  - [ ] Bring RTC integration tests from JupyterLab: https://github.com/jupyterlab/jupyterlab/pull/13240

@fcollonval, am I missing something?

### CODE REVIEW

File: jupyterlab_rtc/_version.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# This file is auto-generated by Hatchling. As such, do not:
+#   - modify
+#   - track in version control e.g. be sure to add to .gitignore
```
Comment (fcollonval) [inline jupyterlab_rtc/_version.py:3]: As mentioned here, this file should not be tracked :wink:

File: lerna.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "lerna": "5.1.0",
+  "version": "independent",
```
Comment (fcollonval) [inline lerna.json:3]: Do we want independent version?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,72 @@
+{
+    "name": "jupyterlab_rtc",
+    "private": true,
+    "version": "4.0.0-alpha.18",
+    "description": "JupyterLab - RTC Extension",
```
Comment (fcollonval) [inline package.json:5]:
```suggestion
    "description": "JupyterLab Extension enabling Real-Time Collaboration",
```

File: packages/docprovider/jest.config.js [hunk 4]
```diff
@@ -3,5 +3,49 @@
  * Distributed under the terms of the Modified BSD License.
  */
 
-const func = require("@jupyterlab/testing/lib/jest-config");
-module.exports = func(__dirname);
+const jestJupyterLab = require('@jupyterlab/testutils/lib/jest-config');
```
Comment (fcollonval) [inline packages/docprovider/jest.config.js:6]: You can simplify this normally - look at: https://github.com/jupyterlab/extension-cookiecutter-ts/blob/4.0/%7B%7Bcookiecutter.python_name%7D%7D/jest.config.js

File: packages/docprovider/tsconfig.json [hunk 5]
```diff
@@ -1,16 +1,9 @@
...
   "include": ["src/*"],
-  "references": [
-    {
-      "path": "../coreutils"
-    },
-    {
-      "path": "../services"
-    }
-  ]
+  "exclude": ["node_modules"]
```
Comment (fcollonval) [inline packages/docprovider/tsconfig.json:8]: This is weird, why did you need that?

File: pyproject.toml [hunk 6]
```diff
@@ -1,75 +1,121 @@
...
-  "Programming Language :: Python :: 3.10"
+    "Intended Audience :: Developers",
+    "Intended Audience :: System Administrators",
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: BSD License",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
```
Comment (fcollonval) [inline pyproject.toml:19]:
```suggestion
    "Programming Language :: Python :: 3.10",
    "Programming Language :: Python :: 3.11",
```

File: pyproject.toml [hunk 7]
```diff
@@ -1,75 +1,121 @@
...
+    "Framework :: Jupyter",
+    "Framework :: Jupyter :: JupyterLab",
+    "Framework :: Jupyter :: JupyterLab :: 3",
```
Comment (fcollonval) [inline pyproject.toml:22]:
```suggestion
    "Framework :: Jupyter :: JupyterLab :: 4",
```

File: scripts/bump_version.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
+#############################################################################
+# Copyright (c) 2018, Voila Contributors                                    #
+# Copyright (c) 2018, QuantStack                                            #
+#                                                                           #
+# Distributed under the terms of the BSD 3-Clause License.                  #
+#                                                                           #
+# The full license is in the file LICENSE, distributed with this software.  #
+#############################################################################
```
Comment (fcollonval) [inline scripts/bump_version.py:8]: Could you update this banner with the Jupyter banner.

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @hbcarlos your plan - including the follow-up PRs - sounds good.

File: pyproject.toml [hunk 9]
```diff
@@ -97,7 +97,8 @@ pydist_resource_paths = ["my-package/img1.png", "my-package/foo/bar.json"]
 [tool.jupyter-releaser.hooks]
 before-build-npm = [
     "python -m pip install 'jupyterlab>=4.0.0a32'",
-    "jlpm"
+    "jlpm",
+    "jlpm build:prod"
```
Comment (fcollonval) [inline pyproject.toml:101]: This is the fix :crossed_fingers:

**Comment by hbcarlos** [comment]
Hi @blink1073, sorry for pinging you, but we can not figure out why the check-release workflow doesn't work.

I'm trying to set up the releaser for this repo, a mono-repo JupyterLab extension with a server extension.
There is an error with the `check-npm` step, saying that some files are missing. See: https://github.com/jupyter-server/jupyter_server_ydoc/actions/runs/4077425029/jobs/7026431557#step:5:1038

The `build-npm` step does not include all the files. It is missing the `lib` and `src` folders. See:
https://github.com/jupyter-server/jupyter_server_ydoc/actions/runs/4077425029/jobs/7026431557#step:5:940

I tried it locally, and it works as expected.

The only weird thing I saw is when building the python package `hatch build`, the tarball `dist/jupyterlab_rtc-4.0.0a18.tar.gz` includes everything except for the folder `packages`. Renaming that folder to something else will fix that.

Do you know what could be wrong?

File: pyproject.toml (hunk 9 above)
Comment (blink1073) [reply pyproject.toml:100]: You also need to run `lerna build` at the top level to produce the `lib` files I believe.

Comment (hbcarlos) [reply pyproject.toml]: I'm running the command `jlpm build:prod` which runs `lerna run build:prod` at the top level but somehow is not building the packages. It says it ran successfully for all the packages but the lib folder isn't there.

Comment (blink1073) [reply pyproject.toml]: You would need a `build:prod` in each of the sub-packages for that to work.  You could run `build` and then `build:prod` I believe, or add `build:prod` scripts to each package.

Comment (hbcarlos) [reply pyproject.toml]: Thanks @blink1073

**Comment by fcollonval** [comment]
@blink1073 I tried to debug this one. The conclusion is that the npm build command is not run inside the releaser copy but in the code repository.

Do you know want may be the cause of that?

**Review by fcollonval — APPROVED** [review]

**Comment by hbcarlos** [comment]
Thanks Fred!
